website/files/models/osfstorage.py

```diff
@@ -102,6 +103,41 @@ def move_under(self, destination_parent, name=None):
             raise exceptions.FileNodeCheckedOutError()
         return super(OsfStorageFileNode, self).move_under(destination_parent, name)
 
+    def check_in_or_out(self, user, checkout, save=False):
+        """
+        Updates self.checkout with the requesting user or None,
+        iff user has permission to check out file or folder.
+        Adds log to self.node.
+
+
+        :param user:        User making the request
+        :param checkout:    Either the same user or None, depending on in/out-checking
+        :param save:        Whether or not to save the user
+        """
+        if (self.is_checked_out and self.checkout != user and 'admin' not in self.node.permissions.get(user._id, []))\
+           or user._id not in self.node.permissions.keys() or 'write' not in self.node.permissions.get(user._id, []):
+            raise exceptions.FileNodeCheckedOutError()
+
+        action = 'checked_out' if checkout else 'checked_in'
+        self.checkout = checkout
+
+        self.node.add_log(
+            action=action,
+            params={
+                'kind': self.kind,
+                'project': self.node.parent_id,
+                'node': self.node._id,
+                'urls': {
+                    'download': "/project/{}/files/{}/{}/?action=download".format(self.node._id, self.provider, self._id),
```

Comment:
Raises `RuntimeError: Attempted to generate a URL without the application context being pushed. This has to be executed when application context is available.`